cpp/src/genetic/reg_stack.cuh

```diff
@@ -58,38 +58,20 @@ struct stack {
    *       to push more than `MaxSize` elements leads to all sorts of incorrect
    *       behavior.
    */
-  HDI void push(DataT val) {
-#pragma unroll
-    for (int i = 0; i < MaxSize; ++i) {
-      if (elements_ == i) {
-        regs_[i] = val;
-        ++elements_;
-      }
-    }
-  }
+  HDI void push(DataT val) { if(!full()) regs_[++elements_] = val; }
 
   /**
-   * @brief Pops the top element from the stack
+   * @brief Lazily pops the top element from the stack
    *
    * @return pops the element and returns it, if already reached bottom, then it
    *         returns zero.
    *
    * @note If called when the stack is already empty, then it just returns a
    *       value of zero! To keep the device-side logic simpler, it has been
    *       designed this way. Trying to pop beyond the bottom of the stack leads
-   *       to all sorts of incorrect behavior.
+   *       to all sorts of incorrect behavior. 
    */
-  HDI DataT pop() {
-#pragma unroll
-    for (int i = 0; i < MaxSize; ++i) {
-      if (elements_ - 1 == i) {
-        --elements_;
-        return regs_[i];
-      }
-    }
-    // shouldn't reach here!
-    return DataT(0);
-  }
+  HDI DataT pop() { return !empty() ? regs_[elements_--] : DataT(0); }
```

Comment:
For the same reasons as above, I think it's better to keep the pop operation the old way... since `maxSize` is not a lot, the penalty due to redundant iterations are lesser than penalty due to register spills...